### DISCUSSION TITLE
c# wrapper

### DISCUSSION THREAD

**Discussion by PVTSoftware** [Macad3D/Macad3D#69] [Q&A]
Hi 
I congratulate you on your software. it's an interesting project to understand c#, geometry and opencascade.
I would like to have more explanations on wrapper generation. 
- Why and how to use CastXml.
-  And the different processes to transform the code before generation (Pass classes)

THANKS

**Comment by dipts** [comment]
Hi,

the WrapperGenerator defines a data model to describe the API of OCCT. The passes work sequentially on this data model to modify it, enhance it or generate information from it. What the individual passes actually do should be largely clear from the passes themselves. 

The task of the first pass is to fill the data model with information from the native OCCT API. To do this, it uses CastXML, which compiles the OCCT header files with Clang/LVVM, extracts the needed information from the AST and transfers it to the pass via an XML file.

In addition to the model and the passes, there are also the TypeWrappers. These define the correct wrapping for various native types in managed types.

**Reply by PVTSoftware** [reply]
Thanks
